### PR TITLE
Bug fixes for BBRAM and unit tests

### PR DESCRIPTION
**Pull request by yperess**
Add some bug fixes for BBRAM drivers and introduce a new generic test suite for read/write operations.

### CODE REVIEW

File: drivers/bbram/bbram_it8xxx2.c [hunk 1]
```diff
@@ -90,13 +111,63 @@ static int bbram_it8xxx2_init(const struct device *dev)
...
+
+	if (offset + count > config->size) {
+		return -ERANGE;
+	}
+
+	bytecpy(buffer, ((uint8_t *)config->base_addr + offset), count);
+	return 0;
+}
+
+static const struct emul_bbram_backend_api it8xxx2_emul_backend_api = {
```
Comment (alevkoy) [inline drivers/bbram/bbram_it8xxx2.c:167]: Would it make more sense for this to be in a separate emulator .c file? I could ask this about some of the other things in this file, but this seemed like the most obvious.

File: drivers/bbram/bbram_microchip_mcp7940n_emul.c [hunk 2]
```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2023 Google LLC
```
Comment (alevkoy) [inline drivers/bbram/bbram_microchip_mcp7940n_emul.c:2]: 2024

File: tests/drivers/bbram/generic/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(bbram)
+
+target_sources(app
+    PRIVATE
+        include/fixture.h
```
Comment (alevkoy) [inline tests/drivers/bbram/generic/CMakeLists.txt:11]: Do you need to list this here? It's not a compilation unit.

File: drivers/bbram/bbram_npcx.c [hunk 4]
```diff
@@ -82,8 +84,9 @@ static int bbram_npcx_write(const struct device *dev, size_t offset, size_t size
 {
 	const struct bbram_npcx_config *config = dev->config;
 
+	printk("offset=%zu, size=%zu, max_size=%zu\n", offset, size, config->size);
```
Comment (ChiHuaL) [inline drivers/bbram/bbram_npcx.c:87]: How about using `LOG_DBG` to replace `printk`?

Comment (yperess) [reply drivers/bbram/bbram_npcx.c:87]: Oh man, good catch, I actually thought I removed all these. Thank you!

File: drivers/bbram/bbram_it8xxx2.c [hunk 5]
```diff
@@ -9,14 +9,26 @@
...
+enum bram_indices {
+	BRAM_IDX_VALID_FLAGS0,
+	BRAM_IDX_VALID_FLAGS1,
+	BRAM_IDX_VALID_FLAGS2,
+	BRAM_IDX_VALID_FLAGS3,
+};
+#endif
 
-LOG_MODULE_REGISTER(bbram, CONFIG_BBRAM_LOG_LEVEL);
+LOG_MODULE_REGISTER(ite_it8xxx2, CONFIG_BBRAM_LOG_LEVEL);
```
[the remark this answers is not in the document]
Comment (yperess) [reply drivers/bbram/bbram_it8xxx2.c:29]: Thanks, reverted.

File: drivers/bbram/bbram_it8xxx2.c (hunk 1 above)
Comment (yperess) [reply drivers/bbram/bbram_it8xxx2.c:167]: Sure, I was trying to keep the change minimal, but you're 100% right that it makes more sense.

File: tests/drivers/bbram/generic/CMakeLists.txt (hunk 3 above)
Comment (yperess) [reply tests/drivers/bbram/generic/CMakeLists.txt:11]: It's a neat trick I learned recently, it makes incremental builds better.

File: subsys/emul/emul.c [hunk 6]
```diff
@@ -55,6 +55,8 @@ int emul_init_for_bus(const struct device *dev)
 		case EMUL_BUS_TYPE_SPI:
 			emul->bus.spi->target = emul;
 			break;
+		case EMUL_BUS_TYPE_NONE:
```
Comment (keith-zephyr) [inline subsys/emul/emul.c:58]: If this is needed here, don't you also need to handle the switch statement at line 68?  Won't you get a warning error in the default case of the 2nd switch?

File: include/zephyr/drivers/bbram/it8xxx2.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	uintptr_t base_addr;
+	/** BBRAM size (Unit:bytes) */
+	int size;
+};
+
+int it8xxx2_emul_backend_set_data(const struct emul *target, size_t offset, size_t count,
+				  const uint8_t *buffer);
+
+int it8xxx2_emul_backend_get_data(const struct emul *target, size_t offset, size_t count,
+				  uint8_t *buffer);
```
Comment (keith-zephyr) [inline include/zephyr/drivers/bbram/it8xxx2.h:27]: These routines don't need to be public if you move the EMUL_DT_INST_DEFINE call into the emulator file.

File: drivers/bbram/bbram_it8xxx2.c [hunk 8]
```diff
@@ -91,12 +105,16 @@ static int bbram_it8xxx2_init(const struct device *dev)
...
 			      &bbram_it8xxx2_driver_api);
 
 DT_INST_FOREACH_STATUS_OKAY(BBRAM_INIT);
+
+#ifdef CONFIG_BBRAM_IT8XXX2_EMUL
+#define BBRAM_EMUL_INIT(inst)                                                                      \
+	EMUL_DT_INST_DEFINE(inst, NULL, NULL, &bbram_cfg_##inst, NULL, &it8xxx2_emul_backend_api)
+
+DT_INST_FOREACH_STATUS_OKAY(BBRAM_EMUL_INIT);
+#endif
```
Comment (keith-zephyr) [inline drivers/bbram/bbram_it8xxx2.c:120]: Does this have to be here?  Can this live in the emul.c file?

File: drivers/bbram/Kconfig.it8xxx2 [hunk 9]
```diff
@@ -8,3 +8,11 @@ config BBRAM_IT8XXX2
 	help
 	  This option enables the BBRAM driver for RISCV_ITE family of
 	  processors.
+
+config BBRAM_IT8XXX2_EMUL
```
Comment (keith-zephyr) [inline drivers/bbram/Kconfig.it8xxx2:12]: Non-blocking, but I would prefer each chipset changes to be split into separate commit.

File: drivers/bbram/bbram_it8xxx2.c [hunk 10]
```diff
@@ -91,12 +105,16 @@ static int bbram_it8xxx2_init(const struct device *dev)
 }
 
 #define BBRAM_INIT(inst)                                                                           \
-	static const struct bbram_it8xxx2_config bbram_cfg_##inst = {                              \
-		.base_addr = DT_INST_REG_ADDR(inst),                                               \
-		.size = DT_INST_REG_SIZE(inst),                                                    \
-	};                                                                                         \
+	BBRAM_DECL_CONFIG(inst);                                                                   \
```
Comment (keith-zephyr) [inline drivers/bbram/bbram_it8xxx2.c:108]: This macro is specific to each chipset.  Please make the names unique.

**Review by alevkoy — DISMISSED** [review]

File: tests/drivers/bbram/generic/CMakeLists.txt (hunk 3 above)
Comment (yperess) [reply tests/drivers/bbram/generic/CMakeLists.txt:11]: Ah, I stand corrected, it wasn't the incremental builds, I misunderstood when it was explained to me (odd since I was noticing a difference in my build environment). Anyway, it allows IDEs to associate the headers with the libraries which gives you better suggestions for the build targets (I use CLion, but my understanding is that this also affects users of Visual Studio).

File: subsys/emul/emul.c (hunk 6 above)
Comment (yperess) [reply subsys/emul/emul.c:58]: Yes, but the function is only called for child nodes of a bus, so this should never run. I needed the extra case to make the compiler happy.

File: include/zephyr/drivers/bbram/it8xxx2.h (hunk 7 above)
Comment (yperess) [reply include/zephyr/drivers/bbram/it8xxx2.h:27]: The emulator declaration sets the same config pointer as the driver. So I needed to declare the device and emulator in the same file. I originally had all of the logic in the same file and these functions as static, @alevkoy requested that I separate the logic which does make it cleaner with this one little exception.

File: drivers/bbram/bbram_it8xxx2.c (hunk 8 above)
Comment (yperess) [reply drivers/bbram/bbram_it8xxx2.c:120]: It needs a pointer to the same `bbram_cfg_##inst`

File: drivers/bbram/bbram_it8xxx2.c (hunk 10 above)
Comment (yperess) [reply drivers/bbram/bbram_it8xxx2.c:108]: Done

File: drivers/bbram/Kconfig.it8xxx2 (hunk 9 above)
Comment (yperess) [reply drivers/bbram/Kconfig.it8xxx2:12]: Done

File: drivers/bbram/bbram_it8xxx2.c (hunk 8 above)
Comment (keith-zephyr) [reply drivers/bbram/bbram_it8xxx2.c:120]: `struct bbram_it8xxx2_config` only has `base_addr` and `size` fields.  Can the emulator just make it's own copy and fill out these values from the devicetree?

File: include/zephyr/drivers/bbram/it8xxx2.h (hunk 7 above)
Comment (keith-zephyr) [reply include/zephyr/drivers/bbram/it8xxx2.h:27]: If this data needs to be shared between the driver and emulator, then you should create a private header file under the drivers/bbram directory.  The include/zephyr directory should only be used for public APIs.

File: drivers/bbram/bbram_it8xxx2.c (hunk 8 above)
Comment (keith-zephyr) [reply drivers/bbram/bbram_it8xxx2.c:120]: Example of another emulator replicated devicetree info:
https://github.com/zephyrproject-rtos/zephyr/blob/b802e42c6f930d94269d41432b2baed74856f4f0/drivers/usb/bc12/emul_bc12_pi3usb9201.c#L318

**Review by alevkoy — DISMISSED** [review]

File: include/zephyr/drivers/bbram/it8xxx2.h (hunk 7 above)
Comment (yperess) [reply include/zephyr/drivers/bbram/it8xxx2.h]: Done

File: drivers/bbram/bbram_it8xxx2.c (hunk 8 above)
Comment (yperess) [reply drivers/bbram/bbram_it8xxx2.c:121]: This works because `pi3usb9201_emul_data_##n` is in the same file, the issue here is that we're actually ignoring the `base_addr` from the devicetree because there's no way to actually know where there will be a free chunk of memory. Instead, when the emulator is enabled we just allocate a static buffer and use that address for the driver. The emulator needs to use the same pointer. I might be able to find this at runtime, but it seemed easier to have them allocated next to each other. The alternative would be to get the `device` pointer when declaring the emulator, then at runtime getting the `conf` pointer. I can change to that if you feel strongly about it.

Comment (alevkoy) [reply drivers/bbram/bbram_it8xxx2.c:121]: I started to leave a comment in the header file wondering if you could expose the buffer as a named object in the emulator and reference it from the device tree to get the address for the device config. It was just going to be a nit, but it would be nice if the driver and emulator could be less coupled.

**Review by alevkoy — DISMISSED** [review]

File: drivers/bbram/bbram_it8xxx2.c (hunk 8 above)
Comment (keith-zephyr) [reply drivers/bbram/bbram_it8xxx2.c:121]: I like Abe's suggestion.  This would still require an `#ifdef CONFIG_BBRAM_IT8XXX2_EMUL` in the real driver to fetch the emulator object, but that seems like a good trade off.

Comment (alevkoy) [reply drivers/bbram/bbram_it8xxx2.c:121]: If the emulator object contained the buffer, is there a way that you could have the emulator DTS node reference the buffer and then have the device node reference that property of the emulator node? Then the "are we using the emulator" logic would be in the .dts file of the build that uses the emulator object.

Comment (yperess) [reply drivers/bbram/bbram_it8xxx2.c]: I think I got as close as I can to what you're looking for. I'm not sure how I would be able to expose the buffer as a named object in the devicetree since there might be multiple instances of it. Let me know if the latest patch works.

**Review by alevkoy — APPROVED** [review]

**Review by keith-zephyr — APPROVED** [review]
